### ISSUE TITLE
unique() does not seem to return correct index if array larger than 16 (migrated from Trac #2063)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3610]
Original ticket http://projects.scipy.org/numpy/ticket/2063
Reported 2012-02-23 by trac user lolowizard, assigned to unknown.

Unique seem to return the index of the last occurence rather than first occurence if len(array)>16

To reproduce:
    v=[0,0,0,0,0,1,1,1,1,1,1,2,2,2,2,2,2]
    w=[0,0,0,0,0,1,1,1,1,1,1,2,2,2,2]
Only difference between v and w is the two extra values at the end.

Then:
    unique(v,return_index=True)
    unique(w,return_index=True)
Look at the results for both commands:
    (array([0, 1, 2]), array([ 0, 10, 15]))

```
(array([0, 1, 2]), array([0, 5, 9]))
```

The two commands should return the same index!

**Comment by thouis** [comment]
Attachment in Trac by trac user lolowizard, 2012-02-23: [python.PNG](http://projects.scipy.org/numpy/attachment/ticket/2063/python.PNG)

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-03

<reformatted description>
you can enclose code in triple curly brackets, no need to attach a png.

Looks like a serious bug, should be fixed for the next release.

**Comment by thouis** [comment]
Comment in Trac by trac user bryan, 2012-03-19

Seems to be a result of argsort not being stable:
    In [34]: np.array(v).argsort()
    Out[34]: array([ 0,  1,  2,  3,  4, 10,  9,  8,  6,  5,  7, 15, 11, 12, 13, 14, 16])

```
In [35]: np.array(w).argsort()
Out[35]: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])
```

**Comment by thouis** [comment]
Comment in Trac by trac user bryan, 2012-03-19

Explicitly specifying mergesort seems to solve the problem:
    In [6]: np.unique(w,return_index=True)
    Out[6]: (array([0, 1, 2]), array([ 0,  5, 11]))

```
In [7]: np.unique(v,return_index=True)
Out[7]: (array([0, 1, 2]), array([ 0,  5, 11]))

index c4441d8..3ecc2d4 100644
--- a/numpy/lib/arraysetops.py
+++ b/numpy/lib/arraysetops.py
atmention:atmention: -174,12 +174,12 atmention:atmention: def unique(ar, return_index=False, return_inverse=False):
             return ar

     if return_inverse or return_index:
-        perm = ar.argsort()
+        perm = ar.argsort(kind='mergesort')
         aux = ar[perm]
         flag = np.concatenate(([True], aux[1:] != aux[:-1]))
         if return_inverse:
             iflag = np.cumsum(flag) - 1
-            iperm = perm.argsort()
+            iperm = perm.argsort(kind='mergesort')
             if return_index:
                 return aux[flag], perm[flag], iflag[iperm]
             else:
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-19

The documentation doesn't seem to imply that the results should be the same, just that they should work. Perhaps this should be raised on the list or made an option.

Union-find would be a fast algorithm for this, but it wouldn't keep the order either, although it might be possible to work around that.

**Comment by thouis** [comment]
Comment in Trac by trac user bryan, 2012-03-19

I am inclined to agree that the documentation does not mention it. OTOH I am not sure what use return_index is, if it doesn't provide some additional guarantees, i.e. that they are the first or last instances. If it's just any index, pretty much all you can use it for is to get one of the unique values, but unique already returns those.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-19

Good point. The documentation should probably be updated also.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-30

Fixed in dbf2351.
